### DISCUSSION TITLE
좋은 API 명세란 무엇일까

### DISCUSSION THREAD

**Discussion by rlajm1203** [JNU-econovation/EEOS-BE#34] [Announcements]
Swagger를 적용하면서, summary와 description을 작성해야 했는데 어떻게 작성해야 할지 감히 잡히지 않았습니다.

BE,FE 각자 개발하기도 바쁜 상황에, Swagger로 작성된 API 명세를 FE 개발자가 한 눈에 이해하지 못하고

BE 개발자와 소통을 해야 하는 상황이라면? 리소스를 들여서 API 명세를 작성한 노력이 헛수고가 되어버립니다.

FE와 BE가 최소한의 소통을 할 수 있도록 하기 위해서는

어떤 식으로 API 명세를 작성해야 할까요? 

API 명세에 어떤 내용이 들어가야 할까요?

@kssumin @Daae-Kim @geongyu09

**Comment by Daae-Kim** [comment]
좋은 api를 고민하며 찾아본 [adobe 테크 블로그 내용](https://blog.developer.adobe.com/three-principles-of-api-first-design-fa6666d9f694)인데요, 

### [api first design 의 3가지 원칙]
**1. Your API is the first user interface of your application**

- api는 우리 서비스를 개발하는 개발자에게 first user interface 이다.
- 그렇기 때문에 이러한 개발자를 고려하여 디자인되어야 한다.

**2. Your API comes first, then the implementation**

- API가 먼저이고, 그다음이 구현이다.
- 애플리케이션이 발전함에 따라 구현부도 발전하고, 최적화되어야 하지만, API는 빈번하게 바뀌어서는 안된다.
- API는 구현에 독립적이어야한다.

**3. Your API is described (and maybe even self-descriptive)**

- API 는 구조화된 설명서이다.
- API를 사용하기 위해서는 API를 작성하지 않은 사람들도 쉽게 이해할 수 있어야 합니다.
- API 그 자체로 설명적인 API를 만들자!

→ 모든 개발자(FE, BE, 우리 서비스를 호출하는 외부개발자)에게 충분히 설명적이며(api를 보고 서비스에 어떤 기능이 있는지 알 수 있는) 구현부의 발전에 독립적인 API가 좋은 API 이다! 라고 말하고 있습니다.

### FE에서는 API 를 어떻게 사용하고 있는지 궁금해서 knotted님와 좋은 api에 대한 대화를 나눠보았습니다.
FE 개발자 knotted 님의 의견으로는,
1. 요청, 반환포맷이 명확해야한다.
2. 파라미터가 왜 필요하고, 언제 필요한지에 대한 설명이 있으면 좋겠다.
3. 어떤 상황에서 어떤 에러가 발생하는지에 대한 내용이 있으면 좋겠다.

ex) (행사 리스트 조회 api) 메인 페이지에서 한페이지에 들어가는 행사리스트를 받아온다. 파라미터는 `category, programStatus, size, page` 가 있고, `category`는 전체 , 주간발표, 행사부가 있으며, `programStatus`의 종류는 A, B, C가 있다. `size`는 a~b 범위를 가진다. `page`에는 ~를 받는다.  이 api 에 대한 `응답`으로 ~가 있고, `에러가 발생하는 경우`는 A, B, C이다. 그때에는 `400 code,  Bad Request`를 반환해준다.
→이렇게 상세한 내용이 필요한 이유는 개발하다가 이것을 왜 썼는지 참고가 필요할때, API 문서를 이용하고, FE 에서는 API 응답값으로 전달되는 데이터를 보고 개발하기때문에 응답값의 변수의 의도를 파악하고 있어야 하기 때문이라고 합니다!

### 위의 내용들을 결합하여 제가 생각하는 좋은 API는,
- api의 요청과 응답이 명시되어 있고, 개발자에게 api 목적과 사용법을 잘 나타내는 것 (로직의 세부사항은 포함하지 않는다.)
이라고 생각합니다!!

**Reply by rlajm1203** [reply]
오! 관련 정보 찾아보시느라 고생 많으셨습니다~!
저도 다애님 글 읽으면서 생각해봤는데, 우리도 EEOS 코드를 물려받은 개발자로서 API 명세가 잘 되어 있다면!
코드를 일일이 까보지 않아도, 얼추 어떻게 동작하는지 파악하기 쉽겠네요!
우리도 BlackCompany 3기분들에게 좋은 환경으로 물려주기 위해서
좋은 API 명세를 하도록 해보죠!

**Reply by rlajm1203** [reply]
knotted님의 의견을 읽어봤을 때, 가장 빠른 API를 만드는 방법은 FE와 BE가 함께 API를 작성하는 것이겠네요,

BE가 먼저 API 설계 초안을 작성하고, 이렇게 작성하게된 이유와 그 사고 방식을 FE에게 잘 설명하는 방식이라는 생각이 듭니다!

예를 들어, 
슬랙 알림기능을 위한 API를 설계할 때 HTTP POST 메소드를 사용하고, RequestBody에 url를 담아줘야 해!
url를 담아서 보내줘야 하는 이유는, 사용자가 접속하는 url은 FE 개발자가 설계하는 것 맞지? (yes)
BE 서버 내에서 url 형식을 저장하고 있으면, FE가 그 설계를 바꾸면 BE에서도 설계를 바꿔줘야 하기 때문에 일을 두 번 하는 일이 없도록
차라리, url 자체를 Request Body에 담아서 보내주는 게 더 효율적일 것 같아! (좋아! 그렇게 하자!)

이렇게 말입니다! (실제로 대화한 내용...)

**Reply by geongyu09** [reply]
좋은 의견 감사합니다! 전적으로 다애님과 종민님의 의견에 동의합니다. 
저 또한 api는 FE, BE가 함께 만들어야 한다는 의견입니다. 이에 추가적인 저의 의견을 추가해 보겠습니다. 

api라는 것은 결국 "인터페이스" 입니다. 
즉 FE, BE는 각자가 각자의 방식대로 로직을 만들어 나가고, 각자의 아키텍처를 구상해 프로덕트를 만들어 갈 텐데,  
그 둘 사이의 통신을 위한 하나의 매개체(== 인터페이스)라고 생각합니다. 
결국 그 누구에게 종속되지 않는 부분이지요! 

그러다보니 어느 한 직책으로 무게감이 실리지 않으면 좋을 것 같습니다! 
논의가 필요한 사항이 생긴다면 함께 논의해나가며, 프로젝트에 온전히 알맞은 api를 작성해 나가야 합니다. 
논의의 경우 하루 날 잡고 한번에 해도 괜찮긴 한데, 
현실적으로 프론트 입장에서는 완벽하게 어떤 api에 정확히 어떤 값이 있어야 하는지 완벽히 파악하기 힘든 경우도 있으며,
기존에 존재하던 api를 재사용 하거나 확장하는 방식으로도 충분히 커버가 가능한 부분이 있을 수 있으나, 온전한 이해가 안 된 상태로 중복되는 api를 만들어버리는 상황이 생길 수도 있을 것 같습니다.
그러니 논의사항이 생기면, 논의할 수 있는 만큼 그때그때 서로 논의하는 시간을 가지는 것이 중요하다고 생각합니다!

물론! FE/BE 를 불문하고 수정사항이 생기거나, 변경해야만 하는 일이 있다면.. 바로바로 이야기를 해주고, 문서에도 기존의 내용에서 어떤 부분이 변경되었는지, 왜 변경하게 되었는지까지 눈에 잘 띄도록 표시해두면 각자 업무를 하면서 실수하는 일이 없어질 것 같습니다! 
당연히 해당 문서가 디벨롭 되는 모습(역사)까지도 볼 수 있겠네요!

정리하자면 제가 생각하는 좋은 문서는 정리를 잘 하는 것(필요한 정보를 모두 표기하는 것. 그로 인해 업무에 지장이 되지 않는 것!)이지만, 이를 잘 작성하기 위해서는 내용적으로 모두가 만족할 수 있어야 하고, 정리 이전에 양쪽의 충분한 논의와 이해가 필요하다고 생각합니다!
